README.md
# firestore_mock

Google Cloud Firestore Test Helpers

## Getting Started

### Create Firestore Mock

```dart
MockFirestore mockFirestore = MockFirestore({
     "articles": {
         "document_id_1": {
           "title": "Flutter is Awesome",
           "author": "uid",
           "views": 123,
         },
         "document_id_2": {
           "title": "Kotlin is better than Dart",
           "author": "uid_2",
           "views": 999,
         }
   }
});
```
### What is implemented (stubbed)?
* Add new document to collection
* Get documents from collection
* Get document by ID from collection
* Query - where, limit, orderBy

Note: Methods (API) that return stream are not stubbed yet.

### Example usage
You can use public cloud_firestore package API.

```dart
final articles = mockFirestore.collection('articles').getDocuments();

print(articles.documents.length); // 3
print(articles.documents[0].documentID); // document_id_1
print(articles.documents[0].data['title']); // Flutter is Awesome
```

CHANGELOG.md
## [0.0.6] 2019/09/26
* Basic implementation of `DocumentReference.snapshots()`

## [0.0.5] 2019/06/19
* Updated cloud_firestore dependency to version 0.11.0+2
* Added support for reading from subcollections

## [0.0.4] - 2019/06/17
* Empty implementation of `DocumentReference.delete()`
* Basic implementation of `DocumentReference.updateData(Map)`

## [0.0.3] - 2019/06/13
* Basic mock implementation for `Query.snapshots()`

## [0.0.2] - 2019/05/03

* `Query.limit`, `Query.orderBy`

## [0.0.1] - 2019/04/30

* Get document from collection
* Add document to collection
* `Query.where(...)` - not working on nested fields yet


test/firestore_mock_test.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firestore_mock/firestore_mock.dart';
import 'package:test/test.dart';

void main() {
  Firestore mockFirestore;

  setUp(() {
    mockFirestore = MockFirestore(data: {
      'articles': {
        'article_id_1': {
          'title': 'Flutter is awesome',
          'author': 'Mr Smith',
          'views': 123,
          'comments': {
            'comment-1': {
              'author': 'Mike Kelly',
              'message': 'Good article.',
            },
            'comment-2': {
              'author': 'Paul Arnold',
              'message': 'Deecent article.'
            },
          }
        },
        'article_id_2': {
          'title': 'Flutter is awesome',
          'author': 'Mr Bean',
          'views': 123,
          'comments': {}
        },
        'article_id_3': {
          'title': 'I love Flutter',
          'author': 'Mr Bean',
          'views': 25,
          'comments': {}
        }
      },
      'users': {
        'uid_1': {'username': 'fakeUser', 'email': 'email@email.com'},
        'uid_2': {'username': 'fakeUser2', 'email': 'fake@email.com'}
      },
      'test_collection': {
        'test_1': {'nullable': null, 'num': 100},
        'test_2': {'nullable': 'non-null', 'num': 90},
        'test_3': {'nullable': null, 'num': 70}
      }
    });
  });

  group('simple CRUD operations', () {
    test('simple fetch data from collection', () async {
      final articlesCollection = mockFirestore.collection('articles');
      final articles = await articlesCollection.getDocuments();

      expect(articles.documents, hasLength(3));
      expect(articles.documents[0].documentID, 'article_id_1');
    });

    test('get document by ID from collection', () async {
      final collection = mockFirestore.collection('users');
      final docReference = collection.document('uid_1');
      final docSnapshot = await docReference.get();

      expect(docSnapshot.documentID, 'uid_1');
      expect(docSnapshot.data,
          {'username': 'fakeUser', 'email': 'email@email.com'});
    });
    test('add - check if added data is the same as provided', () async {
      final userData = {'username': 'newuser', 'email': 'newuser@email.com'};
      final collection = mockFirestore.collection('users');
      final addedUserRef = await collection.add(userData);
      expect(addedUserRef.documentID, isNotNull);

      final addedUserSnap = await addedUserRef.get();
      expect(addedUserSnap.documentID, addedUserRef.documentID);
      expect(addedUserSnap.data, userData);
    });
  });

  test('add - check if added data may be fetched then', () async {
    final userData = {'username': 'newuser', 'email': 'newuser@email.com'};
    final collection = mockFirestore.collection('users');
    final addedUserRef = await collection.add(userData);
    final addedUser = collection.document(addedUserRef.documentID);
    final addedUserSnap = await addedUser.get();

    expect(addedUserSnap.data, userData);
  });

  group('query.where(...)', () {
    test('WHERE equalsTo query', () async {
      final filteredData = await mockFirestore
          .collection('users')
          .where('username', isEqualTo: 'fakeUser')
          .getDocuments();

      expect(filteredData.documents, hasLength(1));
      expect(filteredData.documents[0].documentID, 'uid_1');
      expect(filteredData.documents[0].data,
          {'username': 'fakeUser', 'email': 'email@email.com'});

      final filteredArticles = await mockFirestore
          .collection('articles')
          .where('views', isEqualTo: 123)
          .getDocuments();

      expect(filteredArticles.documents, hasLength(2));
      expect(
        filteredArticles.documents
            .where((snapshot) => snapshot.data['views'] == 123),
        hasLength(2),
      );
    });

    test('WHERE isLessThan query', () async {
      final filteredArticles = await mockFirestore
          .collection('articles')
          .where('views', isLessThan: 100)
          .getDocuments();
      expect(filteredArticles.documents, hasLength(1));
    });

    test('WHERE isLessThanOrEqualTo query', () async {
      final filteredArticles = await mockFirestore
          .collection('articles')
          .where('title', isLessThanOrEqualTo: 'G')
          .getDocuments();
      expect(filteredArticles.documents, hasLength(2));
    });

    test('WHERE isGreaterThan query', () async {
      final filteredArticles = await mockFirestore
          .collection('articles')
          .where('views', isGreaterThan: 25)
          .getDocuments();
      expect(filteredArticles.documents, hasLength(2));
      expect(filteredArticles.documents[0].documentID, 'article_id_1');
    });

    test('WHERE isGreaterThanOrEqualTo query', () async {
      final filteredUsers = await mockFirestore
          .collection('users')
          .where('username', isGreaterThanOrEqualTo: 'fake')
          .getDocuments();
      expect(filteredUsers.documents, hasLength(2));
    });

    test('WHERE isNull query', () async {
      final filteredTests = await mockFirestore
          .collection('test_collection')
          .where('nullable', isNull: true)
          .getDocuments();
      expect(filteredTests.documents, hasLength(2));
      expect(filteredTests.documents[0]['nullable'], isNull);
    });

    test('WHERE combined query', () async {
      final filtered = await mockFirestore
          .collection('test_collection')
          .where('num', isGreaterThanOrEqualTo: 70, isLessThan: 100)
          .getDocuments();
      expect(filtered.documents, hasLength(2));
    });

    test('mulitple call WHERE on query', () async {
      final filtered = await mockFirestore
          .collection('test_collection')
          .where('num', isGreaterThan: 70)
          .where('nullable', isNull: true)
          .getDocuments();

      expect(filtered.documents, hasLength(1));
      expect(filtered.documents[0].data['num'], 100);
    });
  });

  test('query.limit(...) should return limited data', () async {
    final limitedTests = await mockFirestore
        .collection('test_collection')
        .limit(1)
        .getDocuments();

    expect(limitedTests.documents, hasLength(1));
    expect(limitedTests.documents[0].data['num'], 100);
  });

  test('query.orderBy(..., descending = false) should return ordered data',
      () async {
    final orderedTests = await mockFirestore
        .collection('test_collection')
        .orderBy('num')
        .getDocuments();

    expect(orderedTests.documents, hasLength(3));
    expect(orderedTests.documents[0].data['num'], 70);
    expect(orderedTests.documents[1].data['num'], 90);
    expect(orderedTests.documents[2].data['num'], 100);
  });

  test('query.orderBy(..., descending = true) should return ordered data',
      () async {
    final orderedTests = await mockFirestore
        .collection('test_collection')
        .orderBy('num', descending: true)
        .getDocuments();

    expect(orderedTests.documents, hasLength(3));
    expect(orderedTests.documents[0].data['num'], 100);
    expect(orderedTests.documents[1].data['num'], 90);
    expect(orderedTests.documents[2].data['num'], 70);
  });

  test('Query a subcollection', () async {
    final filteredData = await mockFirestore
        .collection('articles')
        .document('article_id_1')
        .collection('comments')
        .getDocuments();

    expect(filteredData.documents, hasLength(2));
    expect(filteredData.documents.elementAt(0).data['author'], 'Mike Kelly');
    expect(filteredData.documents.elementAt(1).data['author'], 'Paul Arnold');
  });

  test('Query a subcollection where', () async {
    final filteredData = await mockFirestore
        .collection('articles')
        .document('article_id_1')
        .collection('comments')
        .where('author', isEqualTo: 'Mike Kelly')
        .getDocuments();

    expect(filteredData.documents, hasLength(1));
    expect(filteredData.documents.elementAt(0).data['author'], 'Mike Kelly');
  });
}


lib/src/utils.dart
import 'dart:math';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:kt_dart/kt.dart';

import 'mock_document.dart';

String randomString(int length) {
  var rand = new Random();
  var codeUnits = new List.generate(length, (index) {
    return rand.nextInt(33) + 89;
  });

  return new String.fromCharCodes(codeUnits);
}

List<String> pathPartsOf(String path) => path.split('/');

List<DocumentReference> collectionDataToDocumentReferences(
  Map<String, Map<String, dynamic>> data,
) {
  return mapFrom(data).map((entry) {
    return MockDocumentReference(documentID: entry.key, data: entry.value);
  }).asList();
}

List<DocumentSnapshot> dataToDocumentSnapshots(
    Map<String, Map<String, dynamic>> data) {
  return collectionDataToDocumentReferences(data)
      .map((ref) => MockDocumentSnapshot.fromRef(ref))
      .toList(growable: false);
}


lib/src/mock_query.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:kt_dart/kt.dart';
import 'package:mockito/mockito.dart';

import '../firestore_mock.dart';
import 'utils.dart';

typedef bool QueryPredicate(KtMapEntry<String, Map<String, dynamic>> entry);

class MockQuerySnapshot extends Mock implements QuerySnapshot {
  final Map<String, Map<String, dynamic>> data;

  MockQuerySnapshot(this.data);

  @override
  List<DocumentSnapshot> get documents => dataToDocumentSnapshots(data);

  @override
  // TODO: implement documentChanges
  List<DocumentChange> get documentChanges => null;
}

class MockQuery extends Mock implements Query {
  final MockFirestore firestore;
  final Map<String, Map<String, dynamic>> data;

  KtMap<String, Map<String, dynamic>> get _ktData => mapFrom(data);

  MockQuery(this.firestore, this.data);

  @override
  Future<QuerySnapshot> getDocuments() => Future.value(MockQuerySnapshot(data));

  @override
  // TODO: Propagate changes
  Stream<QuerySnapshot> snapshots() => Stream.fromFuture(getDocuments());

  @override
  // TODO: WHERE on nested fields
  Query where(String field,
      {isEqualTo,
      isLessThan,
      isLessThanOrEqualTo,
      isGreaterThan,
      isGreaterThanOrEqualTo,
      arrayContains,
      bool isNull}) {
    final filteredData = _ktData
        .filter(_isEqualTo(field, isEqualTo))
        .filter(_isLessThan(field, isLessThan))
        .filter(_isLessThanOrEqualTo(field, isLessThanOrEqualTo))
        .filter(_isGreaterThan(field, isGreaterThan))
        .filter(_isGreaterThanOrEqualTo(field, isGreaterThanOrEqualTo))
        .filter(_arrayContains(field, arrayContains))
        .filter(_isNull(field, isNull))
        .asMap();

    return MockQuery(firestore, filteredData);
  }

  Query limit(int length) {
    if (length >= data.length) {
      return this;
    } else {
      final limitedData =
          _ktData.toList().take(length).associate((pair) => pair).asMap();

      return MockQuery(firestore, limitedData);
    }
  }

  Query orderBy(String field, {bool descending = false}) {
    Map<String, Map<String, dynamic>> sorted;
    if (!descending) {
      sorted = _ktData
          .toList()
          .sortedWith((p1, p2) => p1.second[field].compareTo(p2.second[field]))
          .associate((pair) => pair)
          .asMap();
    } else {
      sorted = _ktData
          .toList()
          .sortedWith((p1, p2) => p2.second[field].compareTo(p2.second[field]))
          .associate((pair) => pair)
          .asMap();
    }

    return MockQuery(firestore, sorted);
  }

  QueryPredicate _isEqualTo(String field, isEqualTo) =>
      (entry) => isEqualTo != null ? entry.value[field] == isEqualTo : true;

  QueryPredicate _isLessThan(String field, isLessThan) => (entry) =>
      isLessThan != null ? entry.value[field].compareTo(isLessThan) < 0 : true;

  QueryPredicate _isLessThanOrEqualTo(String field, isLessThanOrEqualTo) =>
      (entry) => isLessThanOrEqualTo != null
          ? entry.value[field].compareTo(isLessThanOrEqualTo) <= 0
          : true;

  QueryPredicate _isGreaterThan(String field, isGreaterThan) =>
      (entry) => isGreaterThan != null
          ? entry.value[field].compareTo(isGreaterThan) > 0
          : true;

  QueryPredicate _isGreaterThanOrEqualTo(
          String field, isGreaterThanOrEqualTo) =>
      (entry) => isGreaterThanOrEqualTo != null
          ? entry.value[field].compareTo(isGreaterThanOrEqualTo) >= 0
          : true;

  QueryPredicate _arrayContains(String field, arrayContains) =>
      (entry) => arrayContains != null
          ? (entry.value[field] as List).contains(arrayContains)
          : true;

  QueryPredicate _isNull(String field, bool isNull) =>
      (entry) => isNull != null && isNull ? entry.value[field] == null : true;
}


lib/src/mock_document.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firestore_mock/src/mock_collection.dart';
import 'package:meta/meta.dart';
import 'package:mockito/mockito.dart';

class MockDocumentSnapshot extends Mock implements DocumentSnapshot {
  final Map<String, dynamic> data;
  final String documentID;
  final DocumentReference reference;

  MockDocumentSnapshot.fromRef(MockDocumentReference mockReference)
      : data = mockReference.data,
        documentID = mockReference.documentID,
        reference = mockReference;
}

class MockDocumentReference extends Mock implements DocumentReference {
  Map<String, dynamic> data;
  final String documentID;

  MockDocumentReference({@required this.documentID, @required this.data});

  @override
  Future<DocumentSnapshot> get() =>
      Future.value(MockDocumentSnapshot.fromRef(this));

  @override
  Stream<DocumentSnapshot> snapshots() {
    return Stream.fromFuture(get());
  }

  @override
  Future<void> delete() async {
    //TODO: Propagate remove from MockFirestore data
    // Now: Do nothing
  }

  @override
  Future<void> updateData(Map<String, dynamic> data) async {
    this.data = data;
  }

  @override
  CollectionReference collection(String collectionPath) {
    final collectionData = data[collectionPath] as Map<String, dynamic>;

    return MockCollectionReference(
        collection: path, firestore: null, data: collectionData);
  }
}


lib/src/mock_collection.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:meta/meta.dart';

import '../firestore_mock.dart';
import 'mock_document.dart';
import 'mock_query.dart';
import 'utils.dart';

// TODO: Collection changes propagated to snapshots stream
class MockCollectionReference extends MockQuery implements CollectionReference {
  final String collection;

  List<DocumentReference> get documents =>
      collectionDataToDocumentReferences(data);

  MockCollectionReference(
      {@required this.collection,
      @required MockFirestore firestore,
      @required Map<String, Map<String, dynamic>> data})
      : super(firestore, data);

  @override
  DocumentReference document([String path]) =>
      documents.firstWhere((it) => it.documentID == path);

  @override

  /// Creates document with [data] and auto-generated id
  ///
  /// The added document can be fetched then using `firestore.collection(...).document('document_id')`
  Future<DocumentReference> add(Map<String, dynamic> data) {
    final generatedID = randomString(20);
    firestore.data[collection][generatedID] = data;
    this.data[generatedID] = data;

    return Future.value(MockDocumentReference(
      documentID: generatedID,
      data: data,
    ));
  }
}


lib/src/mock_firestore.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:meta/meta.dart';
import 'package:mockito/mockito.dart';

import 'mock_collection.dart';

class MockFirestore extends Mock implements Firestore {
  final Map<String, dynamic> data;

  /// Creates mock for Firestore
  ///
  /// [data] - JSON structured data like in Google Cloud Firestore
  ///
  /// ```json
  /// {
  ///   "articles": {
  ///       "document_id_1": {
  ///         "title": "Flutter is Awesome",
  ///         "author": "uid",
  ///         "views": 123,
  ///       },
  ///       "document_id_2": {
  ///         "title": "Kotlin is better than Dart",
  ///         "author": "uid_2",
  ///         "views": 999,
  ///       }
  /// }
  /// ```
  MockFirestore({@required this.data});

  /// Returns mock for collection at [path].
  ///
  @override
  CollectionReference collection(String path) {
    final collectionData = data[path] as Map<String, dynamic>;

    return MockCollectionReference(
        collection: path, firestore: this, data: collectionData);
  }
}
